source/mir/ndslice/algorithm.d

```diff
@@ -1884,3 +1885,760 @@ size_t countImpl(alias fun, Slices...)(Slices slices)
     while(!slices[0].empty);
     return ret;
 }
+
+private template eachLowerTriangle(alias fun, size_t k = 1,
+                                                bool BeyondMainDiagonal = false)
+{
+    import mir.functional : naryFun;
+    import mir.ndslice.algorithm : eachImpl;
+    import mir.ndslice.slice : Slice, SliceKind;
+
+    static if (__traits(isSame, naryFun!fun, fun))
+    {
+        void eachLowerTriangle(SliceKind kind, Iterator)
+                           (Slice!(kind, [2], Iterator) matrix)
+        {
+            immutable(size_t) m = matrix.length!0;
+            immutable(size_t) n = matrix.length!1;
+
+            size_t i;
+            do
+            {
+                auto e = matrix.front!0;
+
+                static if (!BeyondMainDiagonal)
+                {
+                    if (i >= k)
+                    {
+                        if (i < (n + k))
+                            e[0..(i - k + 1)].eachImpl!fun;
+                        else
+                            e.eachImpl!fun;
+                    }
+                }
+                else
+                {
+                    if (i < (n - k))
+                        e[0..(i + k + 1)].eachImpl!fun;
+                    else
+                        e.eachImpl!fun;
+                }
+                matrix.popFront!0;
+                i++;
+            } while (i < m);
+        }
+    }
+    else
+    {
+        alias eachLowerTriangle = .eachLowerTriangle!(naryFun!fun);
+    }
+}
+
+unittest
+{
+    import mir.ndslice.allocation: slice;
+    import mir.ndslice.topology: iota;
+
+    // 1 2 3
+    // 4 5 6
+    // 7 8 9
+    auto m = iota([3, 3], 1).slice;
+    m.eachLowerTriangle!((ref a) {a = 0; }, 0);
+    assert(m == [
+        [0, 2, 3],
+        [0, 0, 6],
+        [0, 0, 0]]);
+}
+
+unittest
+{
+    import mir.ndslice.allocation: slice;
+    import mir.ndslice.topology: iota, universal;
+
+    // 1 2 3
+    // 4 5 6
+    // 7 8 9
+    auto m = iota([3, 3], 1).slice.universal;
+    m.eachLowerTriangle!((ref a) {a = 0; }, 0);
+    assert(m == [
+        [0, 2, 3],
+        [0, 0, 6],
+        [0, 0, 0]]);
+}
+
+unittest
+{
+    import mir.ndslice.allocation: slice;
+    import mir.ndslice.topology: iota, canonical;
```

Comment:
Is it possible to combine these essentially duplicate tests with an alias seq?
Something like
```
import mir.ndslice.topology: iota, canonical, universal;
/*static*/ foreach(type; AliasSeq!(identity, canonical, universal)
{
    auto m = iota([3, 3], 1).slice.type;
    m.eachLowerTriangle!((ref a) {a = 0; }, 0);
    assert(m == [
        [0, 2, 3],
        [0, 0, 6],
        [0, 0, 0]]);
```

Reply:
Yes, please. I like unittests, but it is nice to keep them smaller with the same coverage.

Reply:
@thewilsonator The way I was able to get it to work looks like
```
version(unittest)
{
    T identity(T)(T x) @property
    {
        return x;
    }
}

unittest
{
    import mir.ndslice.allocation: slice;
    import mir.ndslice.topology: iota, canonical, universal;
    import std.meta: AliasSeq;
    
    void test(alias func)()
    {
        // 1 2 3
        // 4 5 6
        // 7 8 9
        auto m = func(iota([3, 3], 1).slice);
        m.eachUpper!((ref a) {a = 0; }, 0);
        assert(m == [
            [0, 0, 0],
            [4, 0, 0],
            [7, 8, 0]]);
    }
    
    static foreach(type; AliasSeq!(identity, canonical, universal))
    {
        test!type;
    }
}
```